making FLICKR_API_KEY lazy available

caridy opened the pull request:
In case we want to set FLICKR_API_KEY right before calling new Y.App(), this fix the scenario in mojito where the attaching of the modules happens before defining YUI.namespace('Env.Flickr').API_KEY. In the future we might consider passing this key into the model as part of the initial configuration when creating the instance.

caridy committed:
making FLICKR_API_KEY a lazy data in case we want to set it up right before new Y.App(). This fix the scenario in mojito where we the attaching of these modules happens before been able to define YUI.namespace('Env.Flickr').API_KEY. In the future we might consider passing this key into the model as part of the initial configuration when creating the instance.

caridy committed:
grid and lightbox should not explicitely require pnm-photos. they do not use it directly, or create any object/model, they just receive objects as attributes that should fulfill a contract, where grid and lightbox cannot efforce in anyway. This will help to differenciate the responsibilities of each module. At the end of the day, pnm-app is the one responsible for creating and destroying these views and passing the correct data when needed to fulfill the contract between the view and model.

caridy commented:
about f2bc4ef: grid and lightbox should not explicitly require pnm-photos. they do not use it directly, or create any object/model, they just receive objects as attributes that should fulfill a contract, where grid and lightbox cannot efforce in anyway. This will help to differentiate the responsibilities of each module. At the end of the day, pnm-app is the one responsible for creating and destroying these views and passing the correct data when needed to fulfill the contract between the view and model.

ericf commented:
@caridy I wasn't proud with how I ended up using `YUI.Env.FLICKR.API_KEY` to transmit this configuration information around. I think what you have here makes sense. Another option would be to statically set `Y.Photo.FLICKR_API_KEY` directly on the classes which need it.

I don't like the idea of setting this key on each model instance; it would needlessly duplicate the information, and clutter up the model's data attributes with environment configuration.

I agree with f2bc4ef to remove `pnm-photos` from the views' dependencies, good call.

ericf commented on public/js/models/photo.js:
```diff
@@ -14,6 +14,9 @@ Photo = Y.Base.create('photo', Y.Model, [Y.ModelSync.YQL], {
 
     buildQuery: function () {
+        // one time initialization for FLICKR_API_KEY in lazy mode
+        FLICKR_API_KEY = FLICKR_API_KEY || YUI.namespace('Env.Flickr').API_KEY || '';
```
I think it would be cleaner to stick this in a utility function at the top of this file, then anytime it is needed, `getFlickrAPIKey()` can be called.

caridy replied:
Perfecto.

Sent from my iPhone

On May 6, 2012, at 11:31 AM, Eric Ferraiuolo
reply@reply.github.com
wrote:

> > @@ -14,6 +14,9 @@ Photo = Y.Base.create('photo', Y.Model, [Y.ModelSync.YQL], {
> >     pageUrl: 'http://www.flickr.com/photos/{user}/{id}/',
> > 
> > ```
> > buildQuery: function () {
> > ```
> > -        // one time initialization for FLICKR_API_KEY in lazy mode
> > -        FLICKR_API_KEY = FLICKR_API_KEY || YUI.namespace('Env.Flickr').API_KEY || '';
> 
> I think it would be cleaner to stick this in a utility function at the top of this file, then anytime it is needed, `getFlickrAPIKey()` can be called.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ericf/photosnear.me/pull/1/files#r779218

ericf commented on public/js/models/photos.js:
```diff
@@ -20,6 +20,9 @@ Photos = Y.Base.create('photos', Y.ModelList, [Y.ModelSync.YQL], {
         options || (options = {});
 
+        // one time initialization for FLICKR_API_KEY in lazy mode
+        FLICKR_API_KEY = FLICKR_API_KEY || YUI.namespace('Env.Flickr').API_KEY || '';
```
Same as above, this should be moved into a utility fn.

ericf commented on public/js/models/place.js:
```diff
@@ -11,44 +11,34 @@ Place = Y.Base.create('place', Y.Model, [Y.ModelSync.YQL], {
 
     queries: {
-        placeFromId    : 'SELECT {attrs} FROM geo.places WHERE woeid={id}',
-        placeFromText  : 'SELECT {attrs} FROM geo.places WHERE text="{text}"',
```
It looks like this file was out of sync with the changes I recently made to it. This set of changes is also removing a feature that was recently added.

ericf commented:
It looks like you were out of sync when editing `place.js`, your changes also removed a recent feature I added to that file. I'll merge this in using the `--interactive` option to avoid the removal of my recent changes in `place.js`.

caridy commented:
Ops, sorry about that. It was a fresh fork, I thought I was modifying trunk.

Sent from my iPhone

On May 6, 2012, at 11:35 AM, Eric Ferraiuolo
reply@reply.github.com
wrote:

> It looks like you were out of sync when editing `place.js`, your changes also removed a recent feature I added to that file. I'll merge this in using the `--interactive` option to avoid the removal of my recent changes in `place.js`.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/ericf/photosnear.me/pull/1#issuecomment-5538109

ericf commented:
No problem, I can get around it fine.
